File: src/main/scala/com/github/tototoshi/csv/ReaderLineReader.scala
package com.github.tototoshi.csv

import java.io.{ BufferedReader, Reader }

class ReaderLineReader(reader: Reader) extends LineReader {

  private val bufferedReader = new BufferedReader(reader)

  override def readLineWithTerminator(): String = {
    val sb = new StringBuilder
    while ({
      val c = bufferedReader.read()

      if (c == -1) {
        if (sb.isEmpty) return null
        else return sb.toString
      }

      sb.append(c.toChar)

      if (c == '\n' || c == '\u2028' || c == '\u2029' || c == '\u0085') {
        return sb.toString
      }

      if (c == '\r') {
        bufferedReader.mark(1)
        val next = bufferedReader.read()
        if (next == -1) {
          return sb.toString
        } else if (next == '\n') {
          sb.append('\n')
        } else {
          bufferedReader.reset()
        }
        return sb.toString
      }

      true
    }) {}
    sb.toString
  }

  override def close(): Unit = {
    bufferedReader.close()
    reader.close()
  }
}

File: src/main/java/com/github/tototoshi/csv/ReaderLineReader.java
package com.github.tototoshi.csv;

import java.io.IOException;
import java.io.Reader;

/** Splits lines over an internally owned buffer, refilled from the Reader. */
public class ReaderLineReader extends AbstractBufferedLineReader {

    private final Reader reader;

    public ReaderLineReader(Reader reader) {
        this(reader, DEFAULT_BUFFER_SIZE);
    }

    ReaderLineReader(Reader reader, int bufferSize) {
        super(bufferSize);
        this.reader = reader;
    }

    @Override
    protected int fillBuffer(char[] dst) throws IOException {
        return reader.read(dst, 0, dst.length);
    }

    @Override
    public void close() throws IOException {
        reader.close();
    }
}
